src/Compilers/Core/Portable/Emit/EditAndContinue/DeltaMetadataWriter.cs

```diff
@@ -784,6 +812,266 @@ private void PopulateEncLogTableParameters()
+
+            // We need to be able to process parents in the same order as the original metadata for this algorithm to work.
+            var customAttributeParents = _customAttributeParentCounts.Keys.OrderBy(k => CodedIndex.HasCustomAttribute(k)).ToArray();
+
+            // We might need more (for deletes), but we need at least enough rows for each attribute emitted
+            var initialEncLogRowsSize = _customAttributeParentCounts.Sum(t => t.Value);
+
+            // Because we skip emitted rows when processing the original metadata, for efficiency, we can't emit log records
+            // as we go so we use a list to hold the rowIds we will need to log. We need a list because we need direct element
+            // but also need to resize because at this point we can't know how many attributes might have been deleted.
+            var encLogRows = new List<int>(Enumerable.Repeat(0, initialEncLogRowsSize));
+
+            // The first phase of this is to iterate through the (potentially huge) custom attributes from the original
+            // compilation, and issue updates for existing rows as we go.
+            var skippedParents = PopulateEncLogTableCustomAttributesFromOriginalMetadata(customAttributeParents, encLogRows);
+
+            // The second phase is to go through the rows emitted for each delta and issue updates to them, and also
+            // additions for new attributes, from whatever parents were skipped in the first phase
+            PopulateEncLogTableCustomAttributesFromDeltas(skippedParents, encLogRows);
+
+            // Now we just write out our log table that we calculated previously
+            foreach (var rowId in encLogRows)
+            {
+                _customAttributeRows.Add(rowId);
+                metadata.AddEncLogEntry(
+                    entity: MetadataTokens.CustomAttributeHandle(rowId),
+                    code: EditAndContinueOperation.Default);
+            }
+        }
+
+        private List<EntityHandle> PopulateEncLogTableCustomAttributesFromOriginalMetadata(EntityHandle[] emittedAttributeParents, List<int> encLogRows)
+        {
+            var metadataReader = _previousGeneration.OriginalMetadata.MetadataReader;
+
+            // The row id in the original CA table that we're mapping attributes on to
+            var rowId = 1;
+            // The index into emittedAttributeParents, that is the thing we're currently looking for
+            var emittedIndex = 0;
+            // The index in encLogRows we're going to update
+            var encLogRowId = 0;
+
+            // Create a list to store the parents we skip, that come from subsequent generations
+            var skippedParents = new List<EntityHandle>();
+
+            // We iterate through the original metadata, and the emitted attributes at the same time, since they're sorted the same.
+            // metadataReader.CustomAttributes isn't indexable so we're enumerating manually in order to have control of things.
```

Comment:
>// metadataReader.CustomAttributes isn't indexable so we're enumerating manually in order to have control of things. [](http://example.com/codeflow?start=12&length=116)

Technically we can index into it, but having 2 enumerators still makes sense.

Reply:
Binary search was a great idea, the whole thing is much simpler now. Thank you